Persist refresh failure email to all admins

snoe opened the pull request:
When persistence fails to build its cache table for a database we will be sending an email to all admins (OSS+EE) and those with both database detail and monitoring permissions (EE) to alert them of the failure.

### Demo

> **Note**
> Tested on desktop and mobile on Gmail, Apple Mail, Outlook, Spark

> **Warning**
> The demo email renders two errors actually, so there are two similar sections, i.e. it's not a bug 😄 

<details>
<summary>Desktop (Spark email client)</summary>
<img width="937" alt="CleanShot 2022-06-28 at 14 58 38@2x" src="https://user-images.githubusercontent.com/17258145/176197612-9f19a25d-57e1-4f97-885d-0fde9f038973.png">

</details>

<details>
<summary>Mobile (Spark email client)</summary>

![photo_2022-06-28 15 05 21](https://user-images.githubusercontent.com/17258145/176199008-bcd28a9a-fcd4-4e1a-b0bc-798489e8970a.jpeg)


</details>

snoe committed:
Send alert emails on persisted model refresh fail

snoe committed:
Fixing formatting and making sure all the information is available for template

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

dpsutton commented on src/metabase/email/messages.clj:
```diff
@@ -11,6 +11,10 @@
+            [metabase.models :refer [User]]
+            [metabase.models.collection :as collection]
+            [metabase.models.permissions :as perms]
+            [metabase.models.user :as user]
```
refer user from `metabase.models.user :as user :refer [User]` rather than import both. Not a huge deal but `User` in `metabase.models` is just a re-varification of the `User` in models.user.

dpsutton commented on src/metabase/email/messages.clj:
```diff
@@ -246,6 +251,62 @@
+    (when (seq emails)
+      (email/send-message!
+        :subject      (trs "[{0}] Model cache refresh failed for {1}" (app-name-trs) (:name database))
+        :recipients   (vec emails)
```
kinda nothing but what is the purpose of the `vec` call here?

dpsutton replied:
makes me nervous because its either unnecessary or `email/send-message!` is kinda weird and there be dragons but i don't understand them.

snoe replied:
there's a `sequential?` check somewhere deeper, where we have a set of emails here. It is technically a sequential field as you may want a certain person first in the `to` field. Here, we don't care.

dpsutton commented:
lots of failing tests at the moment. And would be nice to include some email screenshots in the PR description. But looks fantastic. nice job.

snoe commented:
> lots of failing tests at the moment. And would be nice to include some email screenshots in the PR description. But looks fantastic. nice job.

This is still in draft until the email is styled, I'll include the screenshot then.

snoe committed:
Fix circular dependency

snoe committed:
Fixing tests and only check advanced perm users if they exist

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

snoe committed:
Fix merge cruft that was renamed

snoe committed:
Only send emails in persist-refresh

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

snoe committed:
Add temporary endpoint for testing persist failure emails

kulyk committed:
Merge branch 'master' into persist-refresh-fail-email

kulyk committed:
Use `border-box` box sizing in emails

kulyk committed:
Use Lato font in emails

kulyk committed:
Style model caching error email

kulyk committed:
Merge branch 'master' into persist-refresh-fail-email

kulyk committed:
Revert "Add temporary endpoint for testing persist failure emails"

This reverts commit 3d3b8f123060f278374aa5a9301c67d547eb42f7.

noahmoss commented on src/metabase/email/_header.mustache:
```diff
@@ -1,21 +1,53 @@
 <head>
+  <link href="https://fonts.googleapis.com/css2?family=Lato:ital,wght@0,100;0,300;0,900;1,100;1,300;1,400;1,700;1,900" rel="stylesheet">
   <style>
+    * {
+      box-sizing: border-box;
+    }
```
https://www.caniemail.com/search/?s=box-sizing

Unfortunately, it's not supported in Outlook on Windows, Yahoo or AOL. 😩 

I think we should definitely strive to support Outlook at minimum. Granted, I don't think our current template or static viz code renders nicely on Outlook, but this would just add an extra issue for anyone trying to debug that.

noahmoss commented on src/metabase/email/messages.clj:
```diff
@@ -246,6 +250,62 @@
+                       :group-by [:pgm.user_id]}
+                      db/query
+                      (mapv :user_id))
+        emails (db/select-field :email User {:where [:and
+                                                     [:= :is_active true]
+                                                     (cond-> [:or
+                                                              [:= :is_superuser true]]
+                                                       (seq user-ids) (conj [:in :id user-ids]))]})
```
This is an edge case, but it's possible for an EE instance to downgrade back to OSS. When this happens, users with DB and/or monitoring perms lose these permissions in the app, since it's an EE-only feature. But we don't run any migrations on the permissions table when downgrading happens, so they'll still show up here.

To fix this you'll need to call `metabase.public-settings.premium-features/enable-advanced-permissions?` to check whether the instance is EE and advanced perms are enabled before including these users in this query.

noahmoss replied:
Also, might be nice to put these queries in a helper function (or augment the existing `all-admin-recipients` function to allow specifying specific feature-level permissions as well).

snoe replied:
Added the checks and tests. Decided against the generic helper function but it's easy enough to write if a use case for a different parameter set comes up.

noahmoss commented on src/metabase/email/persisted-model-error.mustache:
```diff
@@ -0,0 +1,111 @@
+{{> metabase/email/_header }}
+  {{! <style>
+    @media only screen and (max-device-width: 480px) {
+      html, body {
+        font-size: 20px;
+      }
+    }
+  </style> }}
```
nit: should this comment be removed?

noahmoss commented on src/metabase/email/persisted-model-error.mustache:
```diff
@@ -0,0 +1,111 @@
+  </style> }}
+  <div class="card-container">
+    <h2
+      style="margin: 0; padding: 0; font-weight: 700; font-size: 1.5em; color: #4C5773; line-height: 2em;"
```
nit: this color can be referenced from the context as `{{colorTextDark}}`

noahmoss commented on src/metabase/email/persisted-model-error.mustache:
```diff
@@ -0,0 +1,111 @@
+    {{#errors}}
+    <div class="padded-section">
+      <p
+      style="font-weight: 600; font-size: 0.875em; line-height: 1.375em; color: #949AAB;"
```
nit: `{{colorTextMedium}}`

noahmoss replied:
also, seems like outlook has [weird behavior with line-height](https://www.caniemail.com/search/?s=line-height) that requires adding `mso-line-height-rule:exactly` to fix

noahmoss reviewed (changes requested):
BE: requesting small correctness fix in query for recipients. I think a test for that logic would also be good, and/or for the entire email sending flow in general.

Email template: Unfortunately there are a bunch of gotchas where common html/css features are not supported in common email clients like Outlook. (MailDev will render anything, so it's not good for checking compatibility.) I didn't check everything but I commented on some spots that stood out to me as possibly causing issues. 

It might not be possible to avoid all of these issues, but I'd recommend testing out the template in the common mail clients and making sure it doesn't look totally nonsensical, especially since this is a non-trivial template. Eventually it would be nice to use some kind of visual testing service for emails, but we don't have that right now.

kulyk commented on src/metabase/email/_header.mustache:
```diff
@@ -1,21 +1,53 @@
+      margin: auto;
+      background-color: #ffffff;
+    }
+    .padded-section:nth-child(n + 2) {
+      padding-top: 2em;
+    }
```
Thanks for pointing that out Noah. @snoe I need to add some vertical space between error sections if we're going to show more than one error in a single letter. Just to be sure the vertical space is fine, can we pass an extra flag to every error saying if it's not first in the list (e.g. `is-not-first` or something)? I'd prefer error list item indexes, but apparently mustache doesn't support that

snoe replied:
@kulyk will do.

snoe committed:
Add is-not-first to error context to help styling

snoe committed:
Only send to admins if advanced-permissions is off

From @noahmoss:
> This is an edge case, but it's possible for an EE instance to downgrade back to OSS. When this happens, users with DB and/or monitoring perms lose these permissions in the app, since it's an EE-only feature. But we don't run any migrations on the permissions table when downgrading happens, so they'll still show up here.
> To fix this you'll need to call metabase.public-settings.premium-features/enable-advanced-permissions? to check whether the instance is EE and advanced perms are enabled before including these users in this query.

snoe committed:
Test private fn

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-refresh-fail-email

snoe committed:
More robust admin email handling and tests

kulyk committed:
Remove commented out code

kulyk committed:
Avoid `nth-child` selector for padded sections

kulyk committed:
Don't hardcode colors

kulyk committed:
Fix line-height

kulyk committed:
Use line-height fix for Outlook

kulyk committed:
Avoid media-queries and `box-sizing`

kulyk commented on src/metabase/email/_header.mustache:
```diff
@@ -1,21 +1,53 @@
+        padding: 2em 4em;
+        border: 1px solid #dddddd;
+        border-radius: 2px;
+        box-shadow: 0 1px 2px rgba(0, 0, 0, 0.08);
```
I think it won't be a big problem if the shadow doesn't show up somewhere

kulyk commented:
@noahmoss thank you for such thoughtful feedback 🙏 I think I've addressed all of your comments now. I've removed the `box-sizing` prop and tried to move all the critical styles out of media queries. I left some minimal styles there anyway as I can't think of a good way to completely avoid media queries, but this should make sure email will look nice enough if they're not supported. Tested on iOS/macOS for Spark, Outlook Web, Gmail, Apple Mail

noahmoss reviewed (approved):
nice, lgtm

kulyk committed:
Fix E2E test
